File: modules/stego-core/src/main/java/com/shterneregen/securelan/stego/model/ExtractedStegoPayload.java
package com.shterneregen.securelan.stego.model;

import java.nio.charset.StandardCharsets;
import java.util.Objects;

public record ExtractedStegoPayload(StegoContentType contentType, boolean encrypted, byte[] payload) {
    public ExtractedStegoPayload {
        Objects.requireNonNull(contentType, "contentType");
        Objects.requireNonNull(payload, "payload");
        payload = payload.clone();
    }

    @Override
    public byte[] payload() {
        return payload.clone();
    }

    public String asUtf8String() {
        if (contentType != StegoContentType.UTF8_TEXT) {
            throw new IllegalStateException("Extracted payload is not UTF-8 text");
        }
        return new String(payload, StandardCharsets.UTF_8);
    }

    @Override
    public String toString() {
        return "ExtractedStegoPayload[contentType=" + contentType
                + ", encrypted=" + encrypted
                + ", payload=" + payload.length + " bytes]";
    }
}

File: modules/stego-core/src/main/kotlin/com/shterneregen/securelan/stego/model/ExtractedStegoPayload.kt
package com.shterneregen.securelan.stego.model

import java.nio.charset.StandardCharsets
import java.util.Objects

class ExtractedStegoPayload(
    contentType: StegoContentType,
    private val encrypted: Boolean,
    payload: ByteArray,
) {
    private val contentType: StegoContentType = Objects.requireNonNull(contentType, "contentType")
    private val payload: ByteArray = Objects.requireNonNull(payload, "payload").clone()

    fun contentType(): StegoContentType = contentType

    fun encrypted(): Boolean = encrypted

    fun payload(): ByteArray = payload.clone()

    fun asUtf8String(): String {
        if (contentType != StegoContentType.UTF8_TEXT) {
            throw IllegalStateException("Extracted payload is not UTF-8 text")
        }
        return String(payload, StandardCharsets.UTF_8)
    }

    override fun equals(other: Any?): Boolean =
        this === other ||
            (other is ExtractedStegoPayload &&
                contentType == other.contentType &&
                encrypted == other.encrypted &&
                payload == other.payload)

    override fun hashCode(): Int = Objects.hash(contentType, encrypted, payload)

    override fun toString(): String = "ExtractedStegoPayload[contentType=$contentType, encrypted=$encrypted, payload=${payload.size} bytes]"
}
